qualtran/bloqs/data_loading/qroam_clean.py

```diff
@@ -0,0 +1,389 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import numbers
+from collections import defaultdict
+from functools import cached_property
+from typing import cast, Dict, List, Optional, Set, Tuple, Type, TYPE_CHECKING, Union
+
+import attrs
+import numpy as np
+import sympy
+from numpy.typing import ArrayLike
+
+from qualtran import bloq_example, BloqDocSpec, GateWithRegisters, Register, Side, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+from qualtran.bloqs.data_loading.qrom_base import QROMBase
+from qualtran.symbolics import ceil, is_symbolic, log2, prod, SymbolicFloat, SymbolicInt
+
+if TYPE_CHECKING:
+    from qualtran import Bloq, BloqBuilder, SoquetT, QDType
+    from qualtran.simulation.classical_sim import ClassicalValT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+from qualtran.bloqs.data_loading.select_swap_qrom import _alloc_anc_for_reg, SelectSwapQROM
+
+
+def _alloc_anc_for_reg_except_first(
+    bb: 'BloqBuilder', dtype: 'QDType', shape: Tuple[int, ...], dirty: bool
+) -> 'SoquetT':
+    if not shape:
+        return bb.allocate(dtype=dtype, dirty=dirty)
+    soqs = np.empty(shape, dtype=object)
+    ndindex_iter = np.ndindex(shape)
+    _ = next(ndindex_iter)  # Skip the first element.
+    for idx in ndindex_iter:
+        soqs[idx] = bb.allocate(dtype=dtype, dirty=dirty)
+    return soqs
+
+
+def qroam_cost(x, data_size: SymbolicInt, bitsize: SymbolicInt, adjoint: bool = False):
+    # See appendix B of https://arxiv.org/pdf/1902.02134
+    if adjoint:
+        return data_size / x + x
+    else:
+        return data_size / x + bitsize * (x - 1)
+
+
+def get_optimal_log_block_size_clean_ancilla(
+    data_size: SymbolicInt,
+    bitsize: SymbolicInt,
+    adjoint: bool = False,
+    qroam_block_size: Optional[SymbolicInt] = None,
+) -> SymbolicInt:
```

Comment:
docstring.

Reply:
added detailed docstrings to both the classes.